File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-1.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        largest = 0
        stack = []
        n = len(heights)

        for i, h in enumerate(heights):
            while stack and h < heights[stack[-1]]:
                index = stack.pop()
                height = heights[index]

                if stack:
                    width = i - stack[-1] - 1
                else:
                    width = i
                largest = max(largest, width * height)
            stack.append(i)

        while stack:
            index = stack.pop()
            height = heights[index]

            if stack:
                width = n - stack[-1] - 1
            else:
                width = n
            largest = max(largest, width * height)

        return largest

File: Data Structures & Algorithms/max-water-container/submission-1.js
class Solution {
    /**
     * @param {number[]} heights
     * @return {number}
     */
    maxArea(heights) {
        let left = 0;
        let right = heights.length - 1;
        let area = 0;

        while (left < right) {
            let width = right - left;
            let height = Math.min(heights[left], heights[right]);
            area = Math.max(width * height, area);

            // if (heights[left] < heights[right]) {
            //     left++;
            // } else {
            //     right--;
            // }

            if (heights[left] > heights[right]) {
                right--;
            } else {
                left++;
            }
        }

        return area;
    }
}
